onnxruntime/core/providers/coreml/builders/impl/batch_norm_op_builder.cc

```diff
@@ -0,0 +1,139 @@
+  // skip everything except input0 for BatchNormalization
+  const auto& input_defs = node.InputDefs();
+  model_builder.AddInitializerToSkip(input_defs[1]->Name());  // scale
+  model_builder.AddInitializerToSkip(input_defs[2]->Name());  // B
+  model_builder.AddInitializerToSkip(input_defs[3]->Name());  // mean
+  model_builder.AddInitializerToSkip(input_defs[4]->Name());  // var
+}
+
+Status BatchNormalizationOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder,
+                                                          const Node& node,
+                                                          const logging::Logger& /* logger */) const {
+  std::unique_ptr<COREML_SPEC::NeuralNetworkLayer> layer = CreateNNLayer(node);
+
+  const auto& input_defs = node.InputDefs();
+  const auto& initializers(model_builder.GetInitializerTensors());
+  NodeAttrHelper helper(node);
+
+  const auto& scale_tensor = *initializers.at(input_defs[1]->Name());
+  const auto& bias_tensor = *initializers.at(input_defs[2]->Name());
+  const auto& mean_tensor = *initializers.at(input_defs[3]->Name());
+  const auto& var_tensor = *initializers.at(input_defs[4]->Name());
+  const auto eps = helper.Get("epsilon", 1e-5f);
+  const auto channels = scale_tensor.dims()[0];
+
+  auto* coreml_batch_norm = layer->mutable_batchnorm();
+  coreml_batch_norm->set_channels(channels);
+  coreml_batch_norm->set_epsilon(eps);
+  coreml_batch_norm->set_computemeanvar(false);
+  coreml_batch_norm->set_instancenormalization(false);
+
+  CreateCoreMLWeight(*coreml_batch_norm->mutable_gamma(), scale_tensor);   // scale
+  CreateCoreMLWeight(*coreml_batch_norm->mutable_beta(), bias_tensor);     // B
+  CreateCoreMLWeight(*coreml_batch_norm->mutable_mean(), mean_tensor);     // mean
+  CreateCoreMLWeight(*coreml_batch_norm->mutable_variance(), var_tensor);  // var
+
+  *layer->mutable_input()->Add() = node.InputDefs()[0]->Name();
+  *layer->mutable_output()->Add() = node.OutputDefs()[0]->Name();
+
+  model_builder.AddLayer(layer.release());
+  return Status::OK();
+}
+
+// Operator support related
+
+bool BatchNormalizationOpBuilder::IsOpSupportedImpl(const InitializedTensorSet& initializers, const Node& node,
+                                                    const logging::Logger& logger) const {
+  if (node.OutputDefs().size() != 1) {
+    LOGS(logger, VERBOSE) << "Your onnx model may be in training mode, please export "
+                             "it in test mode.";
+    return false;
+  }
+
+  const auto& input_defs = node.InputDefs();
+  std::vector<int64_t> input_shape;
+  if (!GetShape(*input_defs[0], input_shape, logger))
+    return false;
+
+  const auto input_size = input_shape.size();
+  // TODO, support 1d batch normalization (input is 3d)
+  if (input_size != 4) {
+    LOGS(logger, VERBOSE) << "BN only support 4d shape for now, input is "
+                          << input_size << "d shape";
+    return false;
+  }
+
```

Comment:
Can we infer the batch size to be 1 and treat it as 4d if input_size == 3?

Reply:
This will be handled in a future PR, will need to add 1 to the end of the shape and make this a 2d batch norm, such as {N,C,H} -> {N,C,H,1}
Will need to add a ExpandDimsLayer (unsqueeze) before bn and a SqueezeLayer after
This may also be required by other ops since CoreML has weird shape for some layers

---
In reply to: [573438207](https://github.com/microsoft/onnxruntime/pull/6625#discussion_r573438207) [](ancestors = 573438207)